archive/walk.go

```diff
@@ -2,18 +2,232 @@ package archive
+// record timestamps of the corresponding data file. The order of the
+// first & last records matches the archive's data order.
+type seekIndexFile struct {
+	id          ksuid.KSUID
+	first       nano.Ts
+	last        nano.Ts
+	recordCount int64
+}
+
+func (f seekIndexFile) name() string {
+	return fmt.Sprintf("%s-%s-%d-%d-%d.zng", seekIndexKind, f.id, f.recordCount, f.first, f.last)
+}
+
+func (f seekIndexFile) span() nano.Span {
+	return nano.Span{Ts: f.first, Dur: 1}.Union(nano.Span{Ts: f.last, Dur: 1})
+}
+
+var seekIndexNameRegex = regexp.MustCompile(`ts-([0-9A-Za-z]{27})-([0-9]+)-([0-9]+)-([0-9]+).zng$`)
+
+func seekIndexNameMatch(s string) (f seekIndexFile, ok bool) {
+	match := seekIndexNameRegex.FindSubmatch([]byte(s))
+	if match == nil {
+		return
+	}
+	id, err := ksuid.Parse(byteconv.UnsafeString(match[1]))
+	if err != nil {
+		return
+	}
+	recordCount, err := strconv.ParseInt(byteconv.UnsafeString(match[2]), 10, 64)
+	if err != nil {
+		return
+	}
+	firstTs, err := strconv.ParseInt(byteconv.UnsafeString(match[3]), 10, 64)
+	if err != nil {
+		return
+	}
+	lastTs, err := strconv.ParseInt(byteconv.UnsafeString(match[4]), 10, 64)
+	if err != nil {
+		return
+	}
+	return seekIndexFile{
+		id:          id,
+		first:       nano.Ts(firstTs),
+		last:        nano.Ts(lastTs),
+		recordCount: recordCount,
+	}, true
+}
+
+// A tsDir is a directory found in the "<DataPath>/zd" directory of the archive,
+// and holds all of the data & associated files for a span of time, currently
+// fixed to a single day.
+type tsDir struct {
+	nano.Span
+}
+
+func tsDirFor(ts nano.Ts) tsDir {
+	return tsDir{nano.Span{Ts: ts.Midnight(), Dur: nano.Day}}
+}
+
+func parseTsDirName(name string) (tsDir, bool) {
+	t, err := time.Parse("20060102", name)
+	if err != nil {
+		return tsDir{}, false
+	}
+	return tsDirFor(nano.TimeToTs(t)), true
+}
+
+func (t tsDir) name() string {
+	return t.Ts.Time().Format("20060102")
+}
+
+type tsDirVisitor func(tsd tsDir, si []SpanInfo) error
+
+// tsDirVisit calls visitor for each tsDir in the archive, with all spans
+// within the tsDir. tsDir's are visited in the archive's order, and the
+// spans passed to visitor are also sorted by the archive's order.
+func tsDirVisit(ctx context.Context, ark *Archive, filterSpan nano.Span, visitor tsDirVisitor) error {
+	zdDir := ark.DataPath.AppendPath(dataDirname)
+	dirents, err := iosrc.ReadDir(ctx, zdDir)
+	if err != nil {
+		return err
+	}
+	var tsdirs []tsDir
+	for _, e := range dirents {
+		if !e.IsDir() {
+			continue
+		}
+		tsd, ok := parseTsDirName(e.Name())
+		if !ok || !tsd.Overlaps(filterSpan) {
+			continue
+		}
+		tsdirs = append(tsdirs, tsd)
+	}
+	sort.Slice(tsdirs, func(i, j int) bool {
+		if ark.DataSortDirection == zbuf.DirTimeForward {
+			return tsdirs[i].Ts < tsdirs[j].Ts
+		}
+		return tsdirs[j].Ts < tsdirs[i].Ts
+	})
+	for _, d := range tsdirs {
+		dirents, err := iosrc.ReadDir(ctx, zdDir.AppendPath(d.name()))
```

Comment:
This is not going to scale well for dense datasets. At some point we'll want to consider either some callback approach that returns chunks of file infos or generate some kind of file tree that has a max number of leaves before a node is split.